Make Dockerfiles OCI compliant

RedTopper opened the pull request:
## Justification

Closes issue #352

This update makes the Dockerfiles OCI compliant, making it easier to use Buildah or other image building techniques that require it

## Implementation

This changes a few things, listed below:

* auto: Download container is switched to alpine. The `git` container specified the `/git` directory as a volume. As such, all the files under `/git` would be lost after each script invoke. Alpine is used later in the build process anyway, so it shouldn't be any extra cost to switch to it
* auto: "New" clone.sh script is copied into the container, which is basically just the previous clone script that was embedded in the Dockerfile. 
* all: `<<EOF` heredoc styles have been switched to `&& \`
* all: I added NVIDIA_DRIVER_CAPABILITIES and NVIDIA_VISIBLE_DEVICES to expose my Nvidia card. This is most likely a selinux/podman problem, but shouldn't change anything with docker to add it.
* docker-compose: I added selinux labeling. I tested this with real docker (not just podman!) and it seems to work fine. Though I suggest you try it too.

## Testing

Locally builds with buildah. 

Note: for caching to work properly, you still need to replace `/root/.cache/pip` with `/root/.cache/pip,Z` on selinux systems.

Note: I was having some trouble running invoke. Thought it was this PR, but it's a known issue. See https://github.com/invoke-ai/InvokeAI/issues/3182

AbdBarho commented on docker-compose.yml:
```diff
@@ -7,6 +7,9 @@ x-base_service: &base_service
       - &v1 ./data:/data
       - &v2 ./output:/output
     stop_signal: SIGINT
+    security_opt:
+      - label=type:nvidia_container_t
+    runtime: nvidia
```
what would happen here if the machine does not have a gpu?

RedTopper replied:
Fantastic question! I have another machine I can test with that doesn't have an nvidia gpu, I'll report back later today when I get some time to test. My initial thought is to move this statement to the gpu only ones if it explodes.

RedTopper replied:
@AbdBarho Yes, thank you for having me check this! Podman on my non-nvidia tablet complains that it's missing a security context label, so it shouldn't be here by default. Perhaps it's best left commented out? I'm fine with removing it from here and adding it to documentation too.

I didn't try it with Docker on the tablet, but the fact that I'm having issues with Podman makes me think it would also have issues.

AbdBarho replied:
maybe we can use yaml anchors just like the base service.

```yml
x-gpu-config: &gpu_config
   security_opt:
      - label=type:nvidia_container_t
    runtime: nvidia


services:
  auto:
    <<: *base_service
    <<: *gpu_config
    # ...
```

I should also test this setup as well.

RedTopper replied:
@AbdBarho I could do that too. I'd have to make auto-cpu be the "base" and then make auto extend it, otherwise both will get the label. Should I comment out the `<<: *gpu_config` and let people with systems that need to un-comment it?

AbdBarho replied:
Unfortunately, many users of this repo are not very tech savy, and I prefer to not have them fiddle around with the files

but I agree what I wrote is incorrect, this label should only exist on the GPU one...

RedTopper replied:
I added the gpu_config. Please do test!

AbdBarho replied:
That does not seem to work, at least on windows:
```
Error response from daemon: Unknown runtime specified nvidia
```

maybe we can move this problem to another issue / pr so we can finish this.

AbdBarho commented on services/AUTOMATIC1111/clone.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+set -Eeuo pipefail
```
can we also add `x` to flags? just like the shell

RedTopper replied:
Yep, will change

AbdBarho commented on services/AUTOMATIC1111/Dockerfile:
```diff
@@ -1,14 +1,10 @@
-FROM alpine/git:2.36.2 as download
+FROM alpine:3.17 as download
 
-SHELL ["/bin/sh", "-ceuxo", "pipefail"]
```
it is not possible to have shell config at all? no other way to define it?

RedTopper replied:
podman will ignore it completely, at least it does for me.

With that said, I think changing the statement to && should cover `e`, docker/podman already print the full RUN command, so that should cover `x`. 

For pipefail and `u`, I think I can only suggest making complicated RUN command a separate script, like the `clone.sh` one.

AbdBarho reviewed (commented):
Thank you!

> auto: Download container is switched to alpine. The git container specified the /git directory as a volume. As such, all the files under /git would be lost after each script invoke. Alpine is used later in the build process anyway, so it shouldn't be any extra cost to switch to it


This behavior is buildah-specific? 


> docker-compose: I added selinux labeling. I tested this with real docker (not just podman!) and it seems to work fine. Though I suggest you try it too.

Will try!

> Note: for caching to work properly, you still need to replace /root/.cache/pip with /root/.cache/pip,Z on selinux systems.

if this works on all other systems, then we can have it as default.

RedTopper commented:
> This behavior is buildah-specific?

It looks like it is. It seems like Docker used to behave this way too, but something might have changed. This is what lead me to this head-scratching bug: https://stackoverflow.com/a/47575197

> if this works on all other systems, then we can have it as default.

It does not, Docker throws a fit on it. I've been using a script to sed/replace the `,Z`s in and out of the files before committing.

RedTopper commented:
Latest commit:

- I went back to using the `git` alpine container, but create a new /repositories directory instead.
- Swapped auto/auto-cpu inheritance and added gpu_config
- Added "x" to list of set parameters in the bash script

AbdBarho commented:
@RedTopper I changed some final stuff so we can merge this:
* reverted docker-compose, does not seem to work across different OSes yet
* remove `#syntax:dockerfile` from the dockerfiles.

AbdBarho reviewed (approved):
Thank you again
